pkg/webhook/plugins/nodeaffinitywithcache/node_affinity_with_cache.go

```diff
@@ -0,0 +1,200 @@
+	preferredRuntimes = nil
+
+	// get the pod specified bound dataset
+	boundDataSetNames := map[string]bool{}
+	for key, value := range pod.Labels {
+		// if it matches, return array with length 2, and the second element is the match
+		matchString := common.LabelAnnotationPodSchedRegex.FindStringSubmatch(key)
+		if len(matchString) == 2 && value == "required" {
+			boundDataSetNames[matchString[1]] = true
+		}
+	}
+
+	// get the preferred and required runtime
+	preferredRuntimes = map[string]base.RuntimeInfoInterface{}
+	requiredRuntimes = map[string]base.RuntimeInfoInterface{}
+
+	// copy instead of use origin runtimeInfos
+	for name, runtimeInfo := range runtimeInfos {
+		preferredRuntimes[name] = runtimeInfo
+	}
+
+	for name := range boundDataSetNames {
+		// dataset has no runtime, logging info instead of return error
+		runtimeInfo := runtimeInfos[name]
+		if runtimeInfo == nil {
+			log.V(1).Info("labeled dataset has no runtime")
+			continue
+		}
+
+		requiredRuntimes[name] = runtimeInfo
+		delete(preferredRuntimes, name)
+	}
+
+	return
+}
+
+func injectRequiredSchedulingTerms(pod *corev1.Pod, runtimeInfos map[string]base.RuntimeInfoInterface) error {
+	requiredSchedulingTerms := []corev1.NodeSelectorTerm{}
+	for _, runtimeInfo := range runtimeInfos {
+		requiredSchedulingTerm, err := getRequiredSchedulingTerms(runtimeInfo)
+		if err != nil {
+			return err
+		}
+		if requiredSchedulingTerms != nil {
+			requiredSchedulingTerms = append(requiredSchedulingTerms, *requiredSchedulingTerm)
+		}
+	}
+	utils.InjectNodeSelectorTerms(requiredSchedulingTerms, pod)
+
+	return nil
+}
+
+func injectPreferredSchedulingTerm(pod *corev1.Pod, runtimeInfos map[string]base.RuntimeInfoInterface) error {
+	var preferredSchedulingTerms []corev1.PreferredSchedulingTerm
+	for _, runtimeInfo := range runtimeInfos {
+		preferredSchedulingTerm, err := getPreferredSchedulingTerm(runtimeInfo)
+		if err != nil {
+			return err
+		}
+		if preferredSchedulingTerm != nil {
+			preferredSchedulingTerms = append(preferredSchedulingTerms, *preferredSchedulingTerm)
+		}
+	}
+	utils.InjectPreferredSchedulingTerms(preferredSchedulingTerms, pod)
+
+	return nil
+}
+
+func getRequiredSchedulingTerms(runtimeInfo base.RuntimeInfoInterface) (requiredSchedulingTerm *corev1.NodeSelectorTerm, err error) {
+	requiredSchedulingTerm = nil
+	if runtimeInfo == nil {
+		err = fmt.Errorf("RuntimeInfo is nil")
+		return
+	}
+
+	requiredSchedulingTerm = &corev1.NodeSelectorTerm{
+		MatchExpressions: []corev1.NodeSelectorRequirement{
+			{
+				Key:      runtimeInfo.GetCommonLabelName(),
+				Operator: corev1.NodeSelectorOpIn,
+				Values:   []string{"true"},
+			},
+		},
+	}
+	return
+}
+
+func getPreferredSchedulingTerm(runtimeInfo base.RuntimeInfoInterface) (preferredSchedulingTerm *corev1.PreferredSchedulingTerm, err error) {
+	preferredSchedulingTerm = nil
+
+	// logging info instead of return error
+	if runtimeInfo == nil {
+		err = fmt.Errorf("RuntimeInfo is nil")
+		return
```

Comment:
Still return the error or just logging if runtimeinfo is nil?

Reply:
fixed, now return the error.